tracing/src/main/java/com/palantir/tracing/Trace.java

```diff
@@ -28,62 +33,209 @@
  * Represents a trace as an ordered list of non-completed spans. Supports adding and removing of spans. This class is
  * not thread-safe and is intended to be used in a thread-local context.
  */
-public final class Trace {
+public abstract class Trace {
 
-    private final Deque<OpenSpan> stack;
-    private final boolean isObservable;
     private final String traceId;
 
-    private Trace(ArrayDeque<OpenSpan> stack, boolean isObservable, String traceId) {
-        checkArgument(!traceId.isEmpty(), "traceId must be non-empty");
-
-        this.stack = stack;
-        this.isObservable = isObservable;
+    private Trace(String traceId) {
+        checkArgument(!Strings.isNullOrEmpty(traceId), "traceId must be non-empty");
         this.traceId = traceId;
     }
 
-    Trace(boolean isObservable, String traceId) {
-        this(new ArrayDeque<>(), isObservable, traceId);
+    /**
+     * Opens a new span for this thread's call trace, labeled with the provided operation and parent span. Only allowed
+     * when the current trace is empty.
+     */
+    final OpenSpan startSpan(String operation, String parentSpanId, SpanType type) {
+        checkState(isEmpty(), "Cannot start a span with explicit parent if the current thread's trace is non-empty");
+        checkArgument(!Strings.isNullOrEmpty(parentSpanId), "parentSpanId must be non-empty");
+        OpenSpan span = OpenSpan.of(operation, Tracers.randomId(), type, Optional.of(parentSpanId));
+        push(span);
+        return span;
     }
 
-    void push(OpenSpan span) {
-        stack.push(span);
+    /**
+     * Opens a new span for this thread's call trace, labeled with the provided operation.
+     * If the return value is not used, prefer {@link #fastStartSpan(String, SpanType)}}.
+     */
+    final OpenSpan startSpan(String operation, SpanType type) {
+        Optional<OpenSpan> prevState = top();
+        final OpenSpan span;
+        // Avoid lambda allocation in hot paths
+        if (prevState.isPresent()) {
+            span = OpenSpan.of(operation, Tracers.randomId(), type, Optional.of(prevState.get().getSpanId()));
+        } else {
+            span = OpenSpan.of(operation, Tracers.randomId(), type, Optional.empty());
+        }
+
+        push(span);
+        return span;
     }
 
-    Optional<OpenSpan> top() {
-        return stack.isEmpty() ? Optional.empty() : Optional.of(stack.peekFirst());
-    }
+    /**
+     * Like {@link #startSpan(String, String, SpanType)}, but does not return an {@link OpenSpan}.
+     */
+    abstract void fastStartSpan(String operation, String parentSpanId, SpanType type);
```

Comment:
perhaps what's warranted here is just a brand new tracing implementation? I don't think continuing to hack the current implementation even further.

I'm also perplexed by why this would be the primary target for perf improvement, we do lots of expensive processing throughout the stack of calls across our services, and while Tracing is a constant overhead it's likely dwarfed by serde costs.

Reply:
> perhaps what's warranted here is just a brand new tracing implementation?

Sure, there are several problems we could solve in a new tracing implementation, I would be happy to help with that project when I have cycles, but it's not something that I can drive at the moment.

> I'm also perplexed by why this would be the primary target for perf improvement, we do lots of expensive processing throughout the stack of calls across our services, and while Tracing is a constant overhead it's likely dwarfed by serde costs.

That is certainly true of the standard conjure service stack, however this change is not targeting problems with conjure services. We have an internal application with unusually high throughput and low latency requirements, built on our service library (for configuration, status, logging, metrics, etc) where interoperability with the standard stack allows us to avoid reinventing these components.